### PR TITLE
 Add weak generator mode to reference implementation and test vectors

### PR DESCRIPTION
**Pull request by jonasnick**
based on #21 

This PR allows more thorough testing through the introduction of a weak generator mode to the reference implementation and the test vectors. Without this PR the test vectors that
1. use a public key not on the curve 
2. use sig[0:32] >= p 
3. use sig[32:64] >= n 
may increase test coverage in terms of lines but if the implementation actually has a bug they would not discover it. This is because it is very hard to even produce a test case that passes a buggy verification algorithm in the above cases and therefore fails the tests. The weak generator mode allows detection of those bugs.

I'm not 100% convinced that this PR is necessary because 1) it adds complexity both to the reference code and the test vectors 2) because it's hard to produce test cases it'd also be also hard to exploit 3) adding a weak generator mode to an implementation is a giant footgun.

### CODE REVIEW

File: bip-schnorr/reference.py [hunk 1]
```diff
@@ -93,18 +93,45 @@ def schnorr_verify(msg, pubkey, sig):
...
+    return point_add(R0, point_mul(P, e))
+
+# This is a copy of the above schnorr_verify function but with the weak
+# generator mode.
+#
+# The weak_gen argument MUST be False unless used for certain test vectors.
+# When it is True, then "G" is not the secp256k1 generator, but it is chosen as
+# R + e*pubkey.  This is useful to test conditions which are otherwise
+# computationally very hard to create.
+def schnorr_verify(msg, pubkey, sig, weak_gen=False):
```
[the remark this answers is not in the document]
Comment (sipa) [reply bip-schnorr/reference.py]: Ugh, this function overwrites the previous one, which only works through the default weak_gen value.

What about abstracting out a function which does full verification, but takes in G as an argument? The actual `schnorr_verify` can then invoke that with the default generator, while a separate `schnorr_verify_weak_generator` can invoke it combined with the weak generator.

**Comment by sipa** [comment]
Needs rebase.

**Comment by gmaxwell** [comment]
I would personally caution anyone using an implementation that doesn't implement this kind of testing as irresponsible for many applications: it's technique which is simply implemented which has proven itself sensitive to real and serious implementation errors that other forms of testing miss.  ... but that doesn't at all mean that it needs to be in the reference implementation.

Based on what we see with secp2561k1 ecdsa I think that it's likely that 99% of other implementations made from the spec will be extremely reckless-- e.g. no tests at all or maybe just a couple test vectors-- and ill-advisable for anyone to use, and that will be true regardless of what's done here.  Consider: for years googling "python ed25519" returned as the first result an implementation that used H(secret) as the nonce-- even though various ed25519 test vectors that tested the recommended nonce generation procedure were easy to find.

In my view the man purpose of the reference is not really to aid third parties in cooking up their own implementation (anyone who tries that who couldn't pull it off from the text alone is probably far from competent enough to produce a safe one),  but to add clarity to the spec in order to aid cryptanalysis and review of the main implementation. Unfortunately, having a spiffy test seems to get in the way of that clarity (I'd hoped it wouldn't...)

It is, however, unfortunate, that lots of people do go reimplement this stuff without a lot of care or testing... yet many of the tests in libsecp256k1 are fairly generic and could be used against other implementations in a fairly copy and paste manner if not for the fact that they're implemented in a tightly coupled way and more would probably use them if they could copy and paste them.

Perhaps the trade-off here could be resolved by simply providing a separate collection of validation tools, which include e.g. more extensive test vectors,  small subgroup tests (with a forked copy of the reference to demonstrate their use),  implementation generic recommended mutation test procedures, etc.

I think in the known cases where various secp256k1 implementations had flaws they've mostly been in their bignum implementations and wouldn't likely get caught with only a few hundred test vectors created in an implementation blind manner. There have been a few cases where other corner cases (like modular inverse of 0 doing unhelpful things) were an issue, but many of those are less exposed for bip-schnorr than they were with ecdsa. In particular, the pubkey commitment takes a useful avenue for efficient test vector construction off the table.
